### ISSUE TITLE
Can I do Spatio-temporal clustering with HDBSCAN ?

### ISSUE THREAD

**Issue by Stephane-Thales** [scikit-learn-contrib/hdbscan#307]
Hi All,

I've a dataset with latitude, longitude and date-time.
Is there a way to do a spatio-temporal clustering that includes the 3 features?
Havresine is accepting only 2 features (lat lon in radians).
I can scale/normalize the 3 features and use Euclidian, but I believe that I will lose the impact of the physical distance between points.

Any recommendation?

Thanks

**Comment by jc-healy** [comment]
The hard bit here is to work out how distance in seconds relates to distance in radians.  Those are fundamentally two totally different scales.  Any density based clustering technique is going to require a consistent distance between your points.  If you had few enough points you could compute the Haversine distance between points and the temporal distance between your points and fold those together via a scaling that you felt was an appropriate between radian and seconds.  Then you could hand that to hdbscan  as a pre-computed distance matrix.  Just remember that your scaling will change your clustering results.

If you wanted to go deeper down the rabbit hole of combining data with different concepts of distance on different sets of it's features then I'd suggest starting off by reading The 2016 paper "Generalized Low Rank Models" by Udell, et all.   Don't be intimidated by the size of the paper it's a very smooth read with lots of examples.

**Comment by gewitterblitz** [comment]
I have a time distance matrix to create a temporal clusterer. What's the best way to chain it with the Euclidean distance based clusterer to obtain spatio-temporal clustering?

**Comment by gewitterblitz** [comment]
Nevermind, I figured it out.

**Comment by Stephane-Thales** [comment]
> Nevermind, I figured it out.

Hi, @gewitterblitz , do you mind sharing code template for the time-distance matrix and your spatio-temporal clustering solution ?

**Comment by gewitterblitz** [comment]
@Stephane-Thales: I tried the following method but I am not sure if it's the best way. 

```
import hdbscan
from scipy.spatial.distance import pdist

# data is of the form ['time','longitude','latitude']

def compute_squared_EDM(X):
    return squareform(pdist(X, metric='euclidean'))

hdb = hdbscan.HDBSCAN(metric='precomputed',cluster_selection_epsilon=20,min_cluster_size=5)
n, m = data.shape
timeDisMat=compute_squared_EDM(data[:,0].reshape(n, 1)) 
yy = hdb.fit(timeDisMat)
yy.cluster_selection_epsilon = 1e-5
yy.min_cluster_size = 5
disMat = compute_squared_EDM(data[:,1:3])
zz = yy.fit(disMat)
```

FWIW: This did not help me much because my data had some points too close to each other in space but distant in time and the hdbscan output still clustered the time-distant points into one cluster. Let me know if you have found something better.

**Comment by gewitterblitz** [comment]
Yes, you are right. The first just gets overwritten instead of acting on top of the first cluster classification. 

I tried your suggestion but still don't get convincing results. It's most likely due to the uneven scaling of the time and distance dimensions. Not sure if there's an easy way to figure out the correct rescaling method.

I tried 'normalization' ((x-x.min)/(x.max-x.min)) and the `StandardScaler` function from sklearn.preprocessing but both of these didn't help much.

**Comment by Stephane-Thales** [comment]
Perhaps the data need to be normalized also if the shapes are too different ?

**Comment by FalianaRanai** [comment]
@gewitterblitz 
Still no solution?
Can I contact you?

**Comment by Harteg** [comment]
I solved this by first regridding my lat/lon coordinates onto a [HEALPix grid](https://github.com/ntessore/healpix?tab=readme-ov-file), which avoids oversampling near the poles and then converting the grid points to Cartesian coordinates (x, y, z) and combining them with time to perform Euclidean clustering in 4D space (x, y, z, time) using HDBSCAN or similar methods. The main drawback is that the computed distances are straight-line rather than great-circle distances. However, for scales around 2500 km the relative error is under 1%, and for clustering purposes what typically matters is the relative distance between neighbouring points, so I find this an ok tradeoff for a fast clustering. You have to be careful with the scaling between time and space dimensions though. I first apply a StandardScaler() on all dimensions and then play around with different scale factors for the time dimension.
